cli/azd/pkg/ext/hooks.go

```diff
@@ -0,0 +1,379 @@
+package ext
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/azure/azure-dev/cli/azd/pkg/exec"
+	"github.com/azure/azure-dev/cli/azd/pkg/input"
+	"github.com/azure/azure-dev/cli/azd/pkg/osutil"
+	"github.com/azure/azure-dev/cli/azd/pkg/output"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools/bash"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools/powershell"
+	"golang.org/x/exp/slices"
+)
+
+// Generic action function that may return an error
+type InvokeFn func() error
+
+// The type of command hooks. Supported values are 'pre' and 'post'
+type HookType string
+
+const (
+	// Executes pre-command hooks
+	HookTypePre HookType = "pre"
+	// Execute post-command hooks
+	HookTypePost HookType = "post"
+)
+
+// CommandHooks enable support to invoke integration scripts before & after commands
+// Scripts can be invoked at the project or service level or
+type CommandHooks struct {
+	commandRunner exec.CommandRunner
+	console       input.Console
+	cwd           string
+	scripts       map[string]*ScriptConfig
+	envVars       []string
+}
+
+// NewCommandHooks creates a new instance of CommandHooks
+func NewCommandHooks(
+	commandRunner exec.CommandRunner,
+	console input.Console,
+	scripts map[string]*ScriptConfig,
+	cwd string,
+	envVars []string,
+) *CommandHooks {
+	if cwd == "" {
+		osWd, err := os.Getwd()
+		if err != nil {
+			panic(err)
+		}
+
+		cwd = osWd
+	}
```

Comment:
Do we ever want this implicit behavior, instead of having the caller decide?

Reply:
We could expose `cwd` in the script config but at the same time I think it could cause confusion on inferring the script paths.  At that point would the script path be relative to the project/service? or relative to the specified `cwd`.  I'll have a look at gh action & azure workflows to see if their is a precedent we should apply.

Reply:
I was thinking along the lines of just internal class design (nothing to do with external user behavior).

On second thought, if we reject empty, we end up having to surface an `error` which is unfriendly. I think the handling logic of `""` to mean current directory makes total sense, and we should just document it.